pyomo/core/base/reference.py

```diff
@@ -518,10 +521,26 @@ def Reference(reference, ctype=_NotSpecified):
         pass
     elif isinstance(reference, Component):
         reference = reference[...]
+    elif isinstance(reference, _GeneralVarData):
+        obj = SimpleVar()
+        obj._constructed = True
+        obj.domain = reference.domain
+        obj._data = {None: reference}
+        return obj
+    elif isinstance(reference, _GeneralConstraintData):
+        obj = SimpleConstraint()
+        obj._constructed = True
+        obj._data = {None: reference}
+        return obj
+    elif isinstance(reference, _BlockData):
+        obj = SimpleBlock()
+        obj._constructed = True
+        obj._data = {None: reference}
+        return obj
```

Comment:
I would like to know that as well

Reply:
You could make this more general with something like
```
elif isinstance(reference, _ComponentData):
   obj = reference.type()()
   obj.construct()
   obj._data = {None: reference}
```

However, this PR is probably not the best solution: you are creating a new Component to hold the `_data` dict, but that component is never added to a model.  I think a better solution to this will be a way to explicitly create an `_IndexedComponent_slice` to the specific `_ComponentData`.  That is closer to what we do for IndexedComponents (the `[...]` notation in the immediately previous line creates a slice).

Reply:
You're suggesting to create an ```_IndexedComponent_slice``` for the Component with all indices fixed? That seems like a better solution. In this case ```_data``` should map a singleton to the ```_ComponentData``` and ```index``` on line 575 should be a singleton set to index the new Component. When I try this, neither ```_ReferenceDict(reference)``` nor ```_ReferenceSet(reference)``` return the objects I want. I'll look into these functions and see if I can modify them appropriately.

The current implementation is creating a new Component to hold the ```_data``` dict as well, right? And the Component is only added to the model if the user does something like

```
m.r = Reference(comp)
```

or uses ```add_component```. Correct me if I'm wrong here.

Reply:
@Robbybp: yes, I think creating an `_IndexedComponent_slice` is the right approach.  However, I don't think that it will work the way things are currently implemented.  The `_IndexedComponent_slice` object was never designed to work with just a single `_ComponentData`.  Supporting that would need logic that is akin to what is needed to support bounded slices - something that I originally explicitly excluded when building that class, but we should probably consider adding (for ordered Sets).

You are right about the current implementation as well - I missed that you are directly returning the component and not just assigning the `reference` variable.  The problem with your approach is that it ignores the ctype argument to the Reference function.